### PR TITLE
MM-19761 -  User autocomplete should match on spaces

### PR DESCRIPTION
**Pull request by allenlai18**
#### Summary

Possible to search user by their first and last name together with a space. 

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/13328
https://mattermost.atlassian.net/browse/MM-19761

----

Screen Shot 2019-12-12 at 8 02 01 AM

### CODE REVIEW

**Comment by lieut-data** [comment]
@allenlai18, thanks for this change! Can you take a look at adding unit and e2e tests to verify this behaviour, especially as it relates to the many potential edge cases here? This will help both developers and QA as part of the subsequent review.

**Comment by allenlai18** [comment]
> @allenlai18, thanks for this change! Can you take a look at adding unit and e2e tests to verify this behaviour, especially as it relates to the many potential edge cases here? This will help both developers and QA as part of the subsequent review.

@lieut-data Yessir. absolutely.  I just added a unit test. What kind of e2e tests and edge cases do you have in mind?

**Comment by lieut-data** [comment]
@allenlai18 -- good questions. I admit I'm not as well versed as others here, but as a quick stab, I'd consider covering something like 
* autocomplete lets me "not" autocomplete after a space, e.g. `@allen is helping us` shouldn't force a match on someone named `Allen Isaacs`.

I'm imagining other cases like this, especially for existing behaviour.

**Comment by adamjclarkson** [comment]
QQ - is there a way to add users with psuedo-realistic first and last names? The /test command adds users with one name, three names, names with just special characters, etc. None of which are helpful for testing.

**Comment by allenlai18** [comment]
> QQ - is there a way to add users with psuedo-realistic first and last names? The /test command adds users with one name, three names, names with just special characters, etc. None of which are helpful for testing.

@jasonblais  can you help with this?

**Comment by jasonblais** [comment]
> QQ - is there a way to add users with psuedo-realistic first and last names? The /test command adds users with one name, three names, names with just special characters, etc. None of which are helpful for testing.

@saturninoabril or @jespino I think one of you built the `/test` command, would you be able to help?

@adamjclarkson in the meantime you could create a few accounts manually though if you need lots, that might be tedious

**Comment by saturninoabril** [comment]
Thanks Jason!

@allenlai18 You may use those generated from `make test-data`.
Screen Shot 2019-12-20 at 10 37 31 PM

**Comment by adamjclarkson** [comment]
@saturninoabril sorry need a little help - how do I execute the `make test-data` command?

**Comment by saturninoabril** [comment]
@adamjclarkson No worries. You may run it in mattermost-server folder - https://github.com/mattermost/mattermost-server/blob/master/Makefile#L281

**Comment by allenlai18** [comment]
@adamjclarkson  were you able to test it?

**Comment by jasonblais** [comment]
@adamjclarkson Thanks for the review! After PM review, propose adding both QA Review (as you did) and Dev Review labels, and requesting reviews from appropriate team members (1 for QA, 2 for Dev, typically for the team involved in the PR) to ensure the review process continues :)

**Comment by adamjclarkson** [comment]
So I need to add labels and request specific reviewers? Any advice on who to pick? Does it rotate or something?

**Comment by jasonblais** [comment]
That's correct, if not already there (sometimes all reviews are requested at once if they're not dependent on one another). 

Points 3 and 4 here gives some guidance on who to pick for dev/QA reviews:
https://developers.mattermost.com/contribute/getting-started/code-review/#if-you-are-a-core-committer-seeking-a-review

**Comment by allenlai18** [comment]
> Tested and looks good. Please add E2E/Cypress tests, thanks

Working on e2e tests

**Comment by allenlai18** [comment]
@saturninoabril   what would be the best way to get the element and confirm that the user is there?
Screen Shot 2020-01-04 at 3 27 01 PM

**Comment by M-ZubairAhmed** [comment]
@allenlai18 have u tried 'findByText' within that container. There are quite a few tests i did that. It this something you are looking for ?

File: e2e/cypress/integration/at_mentions/at_mentions_user_spec.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+describe('Mention user', () => {
+    before(() => {
+        // # Login and go to /
+        cy.apiLogin('user-1');
+        cy.visit('/');
+    });
+
+    it('M19761 autocomplete should match on spaces', () => {
+        cy.get('#post_textbox').type('@Samuel Tuc');
+        cy.get('.mention__fullname').should('have.text', 'Samuel Tucker');
```
Comment (saturninoabril) [inline e2e/cypress/integration/at_mentions/at_mentions_user_spec.js]: This fails in case the test user has nickname set.  As mentioned by @M-ZubairAhmed, you may use `cy.findByText`. I can see tests something like, (feel free to add/modify test cases as you see fit):
```javascript
function verifySuggestionList({input, expected, withoutSuggestion}) {
    var re = new RegExp(expected, 'g');

    cy.get('#post_textbox').clear().type(input);

    if (withoutSuggestion) {
        cy.get('#suggestionList').should('not.exist');
    } else {
        cy.get('#suggestionList').should('be.visible').within(() => {
            cy.findByText(re).should('be.visible');
        });
    }
}

[
    {input: '@samuel.tucker', expected: 'Samuel Tucker'},
    {input: '@samuel', expected: 'Samuel Tucker'},
    {input: '@tucker', expected: 'Samuel Tucker'},
    {input: '@Samuel', expected: 'Samuel Tucker'},
    {input: '@Tucker', expected: 'Samuel Tucker'},
    {input: '@Samuel Tuc', expected: 'Samuel Tucker'},
    {input: '@Samuel Tucker', expected: 'Samuel Tucker'},
    {input: '@Samuel Tucker ', expected: 'Samuel Tucker', withoutSuggestion: true},
].forEach((testCase) => {
    verifySuggestionList(testCase);
});
```

Comment (allenlai18) [reply e2e/cypress/integration/at_mentions/at_mentions_user_spec.js]: very elegant way of doing it. i like it
